### PR TITLE
Issue 733  wavelet trafo

### PR DESCRIPTION
**Pull request by adler-j**
Work in progress. Fixes some tests but not all. Need input on @kohr-h since you were on the last time.

TODO:

- [ ] Fix adjoint of inverse FourierTransform being wrong.
- [ ] Fix corner cases of WaveletTransform.adjoint
- [ ] Add test for and verify correctness of FT.adjoint with odd sized domain. Currently fails.
- [ ] Add test of `FourierTransform.inverse.adjoint` and `FourierTransform.adjoint.inverse`

### CODE REVIEW

File: odl/test/trafos/fourier_test.py [hunk 1]
```diff
@@ -631,6 +631,53 @@ def char_rect(x):
             assert all_almost_equal(ft.adjoint(ft(char_rect)), discr_rect)
 
 
+def _test_adjoint(operator):
+    """Test adjoint of operator satisfies the definition."""
+    x = noise_element(operator.domain)
+    y = noise_element(operator.range)
```
Comment (kohr-h) [inline odl/test/trafos/fourier_test.py]: These are expected to fail. The adjoint is set equal to the inverse, not (necessarily) the adjoint of the discretized operator. So for functions that fill the entire space, FT is expected to be produce errors. The inverse tests would probably also fail for these cases.
But perhaps the choice of setting adjoint = inverse needs a revisit.

Comment (adler-j) [inline odl/test/trafos/fourier_test.py]: Well the tests are wildly off, not like 5% or so. With that said, I changed the phantom to be a cuboid, which should be nicer.

Comment (adler-j) [inline odl/test/trafos/fourier_test.py]: I seem to be getting a factor 2 across the board, which seems to be the same as we saw in the bug report #732

Comment (adler-j) [inline odl/test/trafos/fourier_test.py]: Example:

```python
>           assert AtAxy.real == pytest.approx(AxAy.real, rel=1e-2, abs=1e-4)
E           assert 1.9999999999999998 == 0.989876379296 ± 9.9e-03
```

**Comment by adler-j** [comment]
Tests are now in place and I am "happy" with them, so to say. Would be great if you could throw an eye on how to fix this @kohr-h.

File: odl/test/trafos/fourier_test.py (hunk 1 above)
Comment (kohr-h) [inline odl/test/trafos/fourier_test.py]: > I seem to be getting a factor 2 across the board, which seems to be the same as we saw in the bug report #732

That could come from the roughly halved range in the real-to-complex case. Maybe that case needs a scaling? It's not exactly 2, it's more like `n / (n // 2 + 1)` probably.

Comment (adler-j) [inline odl/test/trafos/fourier_test.py]: `n / (n // 2 + 1)` sounds very feasible, I saw a change when changing the level of discretization points.

**Comment by adler-j** [comment]
Added  a new commit which makes all the fourier tests run happily, yet we lack a test of `FourierTransform.inverse.adjoint`, which still has the wrong scaling, not to mention `FourierTransform.adjoint.inverse`. We need to fix that.

**Comment by adler-j** [comment]
Also, we have a huge problem with halfcomplex transforms and odd sized domains, which I cannot explain.

File: odl/trafos/fourier.py [hunk 2]
```diff
@@ -1365,13 +1366,25 @@ def _call_pyfftw(self, x, out, **kwargs):
         return out
 
     @property
+    def adjoint(self):
+        """The adjoint Fourier transform."""
+        if self.domain.exponent != 2.0 or self.range.exponent != 2.0:
+            raise NotImplementedError(
+                'Adjoint only implemented for exponent=2')
```
Comment (kohr-h) [inline odl/trafos/fourier.py]: At some point I was considering scrapping these checks. Why should we perform them? In a generalized sense the adjoint relation is still correct even if `p != 2`, so why forbidding it? Just a thought (again).

Comment (ozanoktem) [inline odl/trafos/fourier.py]: L^p is a Hilbert space only for p=2, so the notion of adjoint when 1 < p and p !=2 would have to be based on the Banach space adjoint. I have no idea what happens with these topological issues when considering a discretization, i.e. what happens with the Banach space adjoint in a discretization.

Comment (kohr-h) [inline odl/trafos/fourier.py]: My intuition (somewhat informed since I worked on that stuff in my PhD) is that you get the exact same expressions for the L^p case (1 <= p < infty) since the replacement of the inner product has the same form (integral), and the defining equation for the adjoint is the same.
The only notable difference is that the domain of the operator would be smaller, making it a closed operator only, no longer bounded (for p > 2), and the equality `<F f, g>_(L^q x L^p) = <f, F^* g>_(L^p x L^q)` would only hold for `f, g` such that `f in domain(F)` and `g in domain(F^*)`.

Anyway, regarding warnings I don't know. They are extremely annoying and interfere with screen output a lot. If at all, we would need a method to switch them off.

**Comment by adler-j** [comment]
Further improvement that should be done here (should I do it?) is to change the implementation of the Fourier tests so that we have a fixture that returns `FourierTransform`s that can be used in other tests, i.e. in the tests of the inverse and of the adjoint.

This fixture should also include at least one domain with an odd number of points.

File: odl/trafos/fourier.py (hunk 2 above)
Comment (adler-j) [inline odl/trafos/fourier.py]: My suggestion is to leave this as is and perhaps make a TODO and an issue with this exact discussion quoted. In fact, i just did this: #741.

**Comment by kohr-h** [comment]
> Further improvement that should be done here (should I do it?) is to change the implementation of the Fourier tests so that we have a fixture that returns FourierTransforms that can be used in other tests, i.e. in the tests of the inverse and of the adjoint.
> 
> This fixture should also include at least one domain with an odd number of points.

Good suggestion. I think it makes sense then that we don't work on the same branch simultaneously but push to our respective forks and rebase when necessary.

**Comment by adler-j** [comment]
Pushed the fixtures which adds a few tests and not very surprisingly also causes some failures. This increases the number of tests however, so we may want to move this to the largescale tests (once we get the tests to not fail anymore).

**Comment by adler-j** [comment]
Are you doing any work on your branch with this @kohr-h? Otherwise I would honestly be fine with a partial submission that fixes some of the worst offenders (i.e. at least solves the fourier transform).

**Comment by kohr-h** [comment]
I haven't come around to this lately. The whole thing became a larger refactor on the FT side which is partly done. Forward works but not inverse or adjoint, needs some more massage. I'd say put in partial fixes and perhaps some xfails in the tests where we know things are broken, and we fix those later.

**Comment by adler-j** [comment]
Is this officially on your table now @kohr-h, in that case, do we want to move this to a private branch of yours?

**Comment by kohr-h** [comment]
I'd say no, because I already have [a branch](https://github.com/kohr-h/odl/tree/restructure_ft) on my fork that is based on this and deals with the larger issue of the FT re-organization. This is basically your initial work only. However, I won't have time to look into this in the *near* future. Couple of weeks, maybe.

**Comment by adler-j** [comment]
So we can close this?

**Comment by kohr-h** [comment]
I don't know. Wasn't there some talking about merging the wavelet-specific stuff? I have I just forgotten that this has already been done?
If that's the case then I'd say yes, close this and remove the branch.

**Comment by adler-j** [comment]
The wavelet specific stuff got a quick fix at some point. Closing.
